### PR TITLE
[econstr] Some minor tweaks

### PR DESCRIPTION
**Pull request by ejgallego**
Some minor tweaks removing usage of `Unsafe.to_constr`. They should not matter a lot, but review in needed.

### CODE REVIEW

File: engine/proofview.ml [hunk 1]
```diff
@@ -37,7 +37,7 @@ let proofview p =
 
 let compact el ({ solution } as pv) =
   let nf c = Evarutil.nf_evar solution c in
-  let nf0 c = EConstr.Unsafe.to_constr (Evarutil.nf_evar solution (EConstr.of_constr c)) in
+  let nf0 c = EConstr.Unsafe.to_constr (nf (EConstr.of_constr c)) in
```
Comment (SkySkimmer) [inline engine/proofview.ml]: nf_evar is just safe to_constr @@ of_constr so you can use that to replace the unsafe to_constr.

File: vernac/vernacentries.ml [hunk 2]
```diff
@@ -1586,19 +1586,18 @@ let vernac_check_may_eval ~atts redexp glopt rc =
...
+  let sigma',_ = Evarutil.nf_evars_and_universes sigma' in
   let uctx = Evd.universe_context_set sigma' in
   let env = Environ.push_context_set uctx (Evarutil.nf_env_evar sigma' env) in
-  let c = nf c in
+  let c_ground = EConstr.to_constr sigma' c in
   let j =
-    if Evarutil.has_undefined_evars sigma' (EConstr.of_constr c) then
-      Evarutil.j_nf_evar sigma' (Retyping.get_judgment_of env sigma' (EConstr.of_constr c))
+    if Evarutil.has_undefined_evars sigma' c then
+      Evarutil.j_nf_evar sigma' (Retyping.get_judgment_of env sigma' c)
```
Comment (SkySkimmer) [inline vernac/vernacentries.ml]: Why not the normalized c?

Comment (ejgallego) [reply vernac/vernacentries.ml]: Not sure what you mean, but I think `c_ground` should disappear at some point.

Comment (ejgallego) [reply vernac/vernacentries.ml]: Once `Arguments_renaming` is fixed.

File: engine/proofview.ml (hunk 1 above)
Comment (ejgallego) [reply engine/proofview.ml]: Addressed, thanks!

File: vernac/vernacentries.ml (hunk 2 above)
Comment (SkySkimmer) [reply vernac/vernacentries.ml]: I just feel like we're normalizing multiple times for no reason.
I guess `Eval` and company are not performance critical so it's no a huge deal.

Comment (ejgallego) [reply vernac/vernacentries.ml]: Ok I'll do the tweak and add a comment, anyways this PR need a bench.

**Comment by ejgallego** [comment]
https://ci.inria.fr/coq/job/benchmark-part-of-the-branch/352/

**Comment by ejgallego** [comment]
Hihi, some surprising results here.

**Comment by SkySkimmer** [comment]
Not really, that's all in the noise range.

**Comment by ejgallego** [comment]
I meant in the test suite.

**Comment by SkySkimmer** [comment]
Surprise confirmed.

File: engine/proofview.ml [hunk 3]
```diff
@@ -1211,7 +1211,7 @@ module V82 = struct
    { Evd.it = comb ; sigma = solution }
 
   let top_goals initial { solution=solution; } =
-    let goals = CList.map (fun (t,_) -> fst (Constr.destEvar (EConstr.Unsafe.to_constr t))) initial in
+    let goals = CList.map (fun (t,_) -> fst (EConstr.destEvar solution t)) initial in
```
Comment (SkySkimmer) [inline engine/proofview.ml]: It looks like this change causes a destKO when calling Guard and there is no current goal (all goals solved, or current bullet solved and we're waiting to refocus).

**Comment by maximedenes** [comment]
> Depends on the other EConstr-cleanup PR.

@ejgallego what PR is that?

**Comment by maximedenes** [comment]
@ejgallego Is this on your schedule before the freeze on Thursday?

**Comment by ejgallego** [comment]
> @ejgallego what PR is that?

Already merged.

> @ejgallego Is this on your schedule before the freeze on Thursday?

This needs time to be investigated. As I mentioned in the chat this could even be fixing a bug.

**Comment by maximedenes** [comment]
> This needs time to be investigated. As I mentioned in the chat this could even be fixing a bug.

Is that a yes or a no? :)

**Comment by ejgallego** [comment]
> Is that a yes or a no? :)

That is a I don't know.

**Comment by ejgallego** [comment]
But feel free to push to 8.9, this is not critical in any way.

**Comment by maximedenes** [comment]
> That is a I don't know.

But I'm not asking if you will end up having time for it, I'm asking if it is on your schedule as of now.

**Review by ppedrot — APPROVED** [review]
LGTM, mostly.

**Comment by ejgallego** [comment]
Note the bug that this triggers / introduces and we don't know how to fix @ppedrot

File: engine/proofview.ml [hunk 4]
```diff
@@ -1280,7 +1280,7 @@ module V82 = struct
    { Evd.it = List.map drop_state comb ; sigma = solution }
 
   let top_goals initial { solution=solution; } =
-    let goals = CList.map (fun (t,_) -> fst (Constr.destEvar (EConstr.Unsafe.to_constr t))) initial in
+    let goals = CList.map (fun (t,_) -> fst (EConstr.destEvar solution t)) initial in
```
Comment (SkySkimmer) [inline engine/proofview.ml]: Just revert this change, this function is unsafe and can be looked at in another PR.

Comment (ejgallego) [reply engine/proofview.ml]: Done

**Comment by ppedrot** [comment]
@ejgallego So we can merge this PR, right? The CI failures look spurious.

**Comment by ppedrot** [comment]
Looks like it, I'll merge.
